src/gp_engine.cpp

```diff
@@ -49,39 +88,31 @@ void GPEngine::operator()(const GPOperations::Multiply& op) {
 }
 
 void GPEngine::operator()(const GPOperations::Likelihood& op) {
-  log_likelihoods_(op.dest_idx) = LogLikelihood(op.src1_idx, op.src2_idx);
-}
-
-void GPEngine::operator()(const GPOperations::EvolveRootward& op) {
-  SetTransitionMatrixToHaveBranchLength(branch_lengths_(op.branch_length_idx));
-  plvs_.at(op.dest_idx) = transition_matrix_ * plvs_.at(op.src_idx);
+  SetTransitionMatrixToHaveBranchLength(branch_lengths_(op.dest_idx));
+  auto result = plvs_.at(op.parent_idx).transpose() *
+                                  (transition_matrix_ * plvs_.at(op.child_idx));
+  per_pattern_log_likelihoods_ = result.diagonal().array().log();
+  log_likelihoods_[op.dest_idx] = log(q_[op.dest_idx]) + per_pattern_log_likelihoods_.dot(site_pattern_weights_);
 }
 
-void GPEngine::operator()(const GPOperations::EvolveLeafward& op) {
-  SetTransitionMatrixToHaveBranchLengthAndTranspose(
-      branch_lengths_(op.branch_length_idx));
-  plvs_.at(op.dest_idx) = transition_matrix_ * plvs_.at(op.src_idx);
-}
-
-void GPEngine::operator()(const GPOperations::OptimizeRootward& op) {
-  auto starting_branch_length = branch_lengths_(op.branch_length_idx);
-  std::cout << "starting branch length: " << starting_branch_length << std::endl;
+void GPEngine::operator()(const GPOperations::OptimizeBranchLength& op) {
+  //BrentOptimization(op);
   GradientAscentOptimization(op);
-  std::cout << "after gradient ascent: " << branch_lengths_(op.branch_length_idx)
-            << std::endl;
-  branch_lengths_(op.branch_length_idx) = starting_branch_length;
-  std::cout << "after reset: " << branch_lengths_(op.branch_length_idx) << std::endl;
-  BrentOptimization(op);
-  std::cout << "after brent: " << branch_lengths_(op.branch_length_idx) << std::endl
-            << std::endl;
-}
-
-void GPEngine::operator()(const GPOperations::OptimizeLeafward& op) {
-  Failwith("OptimizeRootward unimplemented for now.");
 }
 
 void GPEngine::operator()(const GPOperations::UpdateSBNProbabilities& op) {
-  Failwith("UpdateSBNProbabilities unimplemented for now.");
+  //Failwith("UpdateSBNProbabilities unimplemented for now.");
+  size_t range_length = op.stop_idx - op.start_idx;
+  if (range_length == 1) {
+    q_(op.start_idx) = 1.;
+    return;
+  }
+
+  // Assumption: log_likelihoods_ have been updated on [op.start_idx, op.stop_idx).
```

Comment:
Let's put this documentation somewhere more prominent.

Reply:
Any suggestion on the best place for it to go?